readme
	
	Nintendo Entertainment System (NES) loader module
	------------------------------------------------------
	Copyright 2006, Dennis Elser (dennis@backtrace.de)


	IDA Pro loader module for Nintendo Enternainment System
  (NES) ROM images in iNES file format.

  You need IDA Pro 4.9 at least in order to be able to
  use this loader. For installation instructions, please
  read below.

  This loader tells IDA Pro how to correctly load
	iNES ROM images. The disassembly is done by the
	M6502 processor module coming along with IDA Pro,
  which is compatible with the slightly modified
  cpu of the NES.
	
	Please note that the real NES hardware memory area
	ends after 0xFFFF as it can address 16bit only.
	The NES uses several page/bank swapping techniques in
  order to load additional ROM banks. However, this
  loader reads at most two 16k PRG ROM banks into the
  IDA database, because it takes care of the original
  memory layout.
	This significantly improves the disassembly, but currently
  doesn't allow to reassemble the disassembly.

  By the way: if you want the IDA wizard to show up
  the NES loader under the "Consoles" tab, you can edit
  the following files:
  
    file: %idadir%/cfg/templates.xml
    add the following lines below under the "Consoles" category:
    
      <template idc:icon="I_ROM" extensions="ROM_NES" loader="nes" processor="m6502">
        <xi:include xpointer="xpointer(/ida/templates/wizard[@X='default_w'])"/>
      </template>
    
    file: %idadir%/cfg/idagui.cfg
    add the following line below under "FILE_EXTENSIONS":
      
      ROM_NES, ".nes Nintendo NES ROM Images", "*.nes"
      
      
	Copy compiled loader to %idadir%/loaders/nes.ldw
	
	-------------------------------------------------------------------      
      
  Note to programmers:
  
  The whole file is stored within the IDA Pro database
  by the loader. You can access all parts of the original
  file using netnodes.
  
  Example:
  
  The original iNES header is stored in a netnode
  (INES_HDR_NODE). You can access the header from a
  plugin by reading the netnode's blob as shown below.
	

	Example code:

		#include "nes.h"

		netnode node(INES_HDR_NODE);
		ines_hdr hdr;

		node.getblob(&hdr, &INES_HDR_SIZE, 0, 'I');
      
      
      
  (c) 2006, Dennis Elser      

history
	history:
	--------

        2006-Sep-28 version 0.24b - beta 1 build 3
        -------------------------
        - SRAM segment is now always created. I have partly
          reverse engineered a game which uses SRAM to store
          some flags, although there is SRAM available
          in the cartridge. (Maybe I should grab some more up
          to date documents on the NES?)


        2006-Sep-22 version 0.24b - beta 1 build 2
        -------------------------
        - inf.filetype is not set anymore (deprecated).
          get_loader_name() will be used instead for a
          possible bank switching plugin to determine the
          filetype
          

        2006-Sep-21 version 0.24b - beta 1
        -------------------------
        - when a corrupted NES image is detected, a dialog box
          is shown, asking the user whether to fix the header or
          not
        - store_chr_rom_pages() and store_prg_rom_pages() are fixed
        - sourcecode restructured / added prototypes for every function
        - added convenience functions, making the source more readable
        - renamed all "store_..." functions to "save_..._as_blob()"
        - fixed a memory leak in save_trainer_as_blob()
        - save_ines_hdr_as_blob( ines_hdr *hdr) changed to
          save_ines_hdr_as_blob( void )
        - load_chr_rom_bank() is more verbose (reminder ;)
        - load_trainer() is more verbose
        - entrypoint is set to RESET_VECTOR_START_ADDRESS
        - moved mapper enum and char *mapper_names[] to mappers.h
        

        2006-Sep-19 version 0.23b - beta 1
        -------------------------
        - fixed a bug in get_vector()
        - removed vector_t
        - removed v_ constants from nes.h
        - vector comments are removed
        - GNROM mappers don't seem to load the first two
          PRG pages, they now load the first and the last.
          I probably should get a more up to date document
          on mappers
        - segments are now patchable, this means the file
          offset is shown below IDA's disassembly as well
        - added fix_ines_hdr()
        - added mapper names and helper function get_mapper_name()
        

        2006-Sep-18 version 0.21b
        -------------------------
        - added PPU RAM layout to nes.h
        - fixed bug in load_chr_rom_bank()
        - temporarily "disabled" load_chr_rom_bank(), PPU is
          not fully supported, yet
        - renamed some functions
        - commented the code a bit more


        2006-Sep-15 version 0.20b
        -------------------------
        - removed load_prg_rom_pages() and load_chr_rom_pages()
        - added load_prg_rom_bank() and load_chr_rom_bank()
        - added identifiers for some mappers
        - began adding support for known mappers by
          implementing load_rom_image(), which is a wrapper
          for load_prg_rom_bank() and load_chr_rom_bank().
          ROM images with known mappers should now correctly
          be loaded
          

        2006-Sep-14 version 0.17b
        -------------------------
        - source code has been restructured
        - all relevant NES/ROM segments are now created,
          but not all segments are being initialized.
          expansion rom segment is always created.
        - ioregs.h has been added to the project
        - I/O registers are now named and commented,
          the disassembly is now pretty readable!
        - added store_trainer(), store_prg_rom_pages()
          and store_chr_rom_pages() - those functions
          store (together with the already available
          store_ines_hdr() function) the whole original
          file inside the IDA database. This allows virtual
          bank/page swapping as well as output file generation
          from an external plugin.


        2006-Sep-12 version 0.16b
        -------------------------
        - vectors are now commented with their respective file offsets.
          This is useful for "hooking" vectors (trainer development etc.)


        2006-Sep-11 version 0.15b
        -------------------------
        - trainers are supported (tested on my own trainer
          that I've coded for a game today - it patches
          the NMI vector to jump to $7000 - where trainers are mapped to.
          I don't know about other trainers, since I haven't seen
          any trainers "in the wild".


        2006-Sep-10 version 0.14b build 2
        -------------------------
        - fixed is_corrupt_ines_hdr()
        - loading of CHR ROM pages is currently limited
          to loading the first CHR ROM page at CHR_ROM_START_ADDRESS
        - added add_entry_points() and name_vectors()
        - added void create_sram_page(), SRAM virtually
          supported, but untested


        2006-Sep-09 version 0.14b
        -------------------------
        - added sanity checks
        - nes.h: added vector_t
        - nes.cpp: renamed get_reset_vector() to
          get_vector()
        - IRQ and NMI routines are now exported as
          entrypoints (public names) as well

    
        2006-Sep-08 version 0.13b
        -------------------------
        - added get_reset_vector() : entrypoint now equals
          the address of the reset vector (stored inside
          the last PRG ROM page?!)
        - load_prg_rom_pages() temporarily loads a maximum
          of two pages - possibly until all mappers are
          supported
        - changed load_prg_rom_pages() and load_chr_rom_pages()
          to temporarily create nonpatchable segments


        2006-Sep-03 version 0.12b
        -------------------------
        - added a check for corrupted iNES headers
        - added a check for presence of a trainer
        - added INES_MASK_... macros to nes.h
        - added dumpidb.idc to package


		    2006-Sep-02 version 0.11b
		    -------------------------
		    - added support for CHR-ROM pages
		    - ROM image is now completely loaded

src/ioregs.h
/*

	Nintendo Entertainment System (NES) loader module
	------------------------------------------------------
	Copyright 2006, Dennis Elser (dennis@backtrace.de)

*/

#ifndef _IOREGS_H
#define _IOREGS_H


// size of 8 and 16 bit I/O registers
#define IOREG_8                             1
#define IOREG_16                            2

// Picture Processing Unit I/O regs
#define PPU_CR_1_ADDRESS                    0x2000
#define PPU_CR_1_SIZE                       IOREG_8
#define PPU_CR_1_SHORT_DESCRIPTION          "PPU_CR_1"
#define PPU_CR_1_COMMENT                    "PPU Control Register #1 (W)\n\n" \
                                            "   D7: Execute NMI on VBlank\n" \
                                            "   D6: PPU Master/Slave Selection\n" \
                                            "   D5: Sprite Size\n" \
                                            "   D4: Background Pattern Table Address\n" \
                                            "   D3: Sprite Pattern Table Address\n" \
                                            "   D2: PPU Address Increment\n" \
                                            "D1-D0: Name Table Address"

#define PPU_CR_2_ADDRESS                    0x2001
#define PPU_CR_2_SIZE                       IOREG_8
#define PPU_CR_2_SHORT_DESCRIPTION          "PPU_CR_2"
#define PPU_CR_2_COMMENT                    "PPU Control Register #2 (W)\n\n" \
                                            "D7-D5: Full Background Colour (when D0 == 1)\n" \
                                            "D7-D5: Colour Intensity (when D0 == 0)\n" \
                                            "   D4: Sprite Visibility\n" \
                                            "   D3: Background Visibility\n" \
                                            "   D2: Sprite Clipping\n" \
                                            "   D1: Background Clipping\n" \
                                            "   D0: Display Type"

#define PPU_SR_ADDRESS                      0x2002
#define PPU_SR_SIZE                         IOREG_8
#define PPU_SR_SHORT_DESCRIPTION            "PPU_SR"
#define PPU_SR_COMMENT                      "PPU Status Register (R)\n\n" \
                                            "   D7: VBlank Occurance\n" \
                                            "   D6: Sprite #0 Occurance\n" \
                                            "   D5: Scanline Sprite Count\n" \
                                            "   D4: VRAM Write Flag\n\n" \
                                            "" \
                                            "NOTE: D7 is set to 0 after read occurs.\n" \
                                            "NOTE: After a read occurs, $2005 is reset, hence the\n" \
                                            "      next write to $2005 will be Horizontal.\n" \
                                            "NOTE: After a read occurs, $2006 is reset, hence the\n" \
                                            "      next write to $2006 will be the high byte portion."

// Sprite RAM I/O regs
#define SPR_RAM_AR_ADDRESS                  0x2003
#define SPR_RAM_AR_SIZE                     IOREG_8
#define SPR_RAM_AR_SHORT_DESCRIPTION        "SPR_RAM_AR"
#define SPR_RAM_AR_COMMENT                  "SPR-RAM Address Register (W)\n\n" \
                                            "D7-D0: 8-bit address in SPR-RAM to access via $2004."

#define SPR_RAM_IOR_ADDRESS                 0x2004
#define SPR_RAM_IOR_SIZE                    IOREG_8
#define SPR_RAM_IOR_SHORT_DESCRIPTION       "SPR_RAM_IOR"
#define SPR_RAM_IOR_COMMENT                 "SPR-RAM I/O Register (W)\n\n" \
                                            "D7-D0: 8-bit data written to SPR-RAM."

// VRAM I/O regs
#define VRAM_AR_1_ADDRESS                   0x2005
#define VRAM_AR_1_SIZE                      IOREG_8
#define VRAM_AR_1_SHORT_DESCRIPTION         "VRAM_AR_1"
#define VRAM_AR_1_COMMENT                   "VRAM Address Register #1 (W2)\n\n" \
                                            "Commonly used used to \"pan/scroll\" the screen (sprites\n" \
                                            "excluded) horizontally and vertically. However, there\n" \
                                            "is no actual panning hardware inside the NES. This\n" \
                                            "register controls VRAM addressing lines."

#define VRAM_AR_2_ADDRESS                   0x2006
#define VRAM_AR_2_SIZE                      IOREG_8
#define VRAM_AR_2_SHORT_DESCRIPTION         "VRAM_AR_2"
#define VRAM_AR_2_COMMENT                   "VRAM Address Register #2 (W2)\n\n" \
                                            "Commonly used to specify the 16-bit address in VRAM to\n" \
                                            "access via $2007. However, this register controls VRAM\n" \
                                            "addressing bits, and therefore should be used with\n" \
                                            "knowledge of how it works, and when it works."


#define VRAM_IOR_ADDRESS                    0x2007
#define VRAM_IOR_SIZE                       IOREG_8
#define VRAM_IOR_SHORT_DESCRIPTION          "VRAM_IOR"
#define VRAM_IOR_COMMENT                    "VRAM I/O Register (RW)\n\n" \
                                            "D7-D0: 8-bit data read/written from/to VRAM."

// Pseudo Audio Processing Unit I/O regs

// pulse 1
#define PAPU_PULSE_1_CR_ADDRESS             0x4000
#define PAPU_PULSE_1_CR_SIZE                IOREG_8
#define PAPU_PULSE_1_CR_SHORT_DESCRIPTION   "pAPU_P_1_CR"
#define PAPU_PULSE_1_CR_COMMENT             "pAPU Pulse #1 Control Register (W)"
                                            
#define PAPU_PULSE_1_RCR_ADDRESS            0x4001
#define PAPU_PULSE_1_RCR_SIZE               IOREG_8
#define PAPU_PULSE_1_RCR_SHORT_DESCRIPTION  "pAPU_P_1_RCR"
#define PAPU_PULSE_1_RCR_COMMENT            "pAPU Pulse #1 Ramp Control Register (W)"
                                            
#define PAPU_PULSE_1_FTR_ADDRESS            0x4002
#define PAPU_PULSE_1_FTR_SIZE               IOREG_8
#define PAPU_PULSE_1_FTR_SHORT_DESCRIPTION  "pAPU_P_1_FTR"
#define PAPU_PULSE_1_FTR_COMMENT            "pAPU Pulse #1 Fine Tune (FT) Register (W)"
                                            
#define PAPU_PULSE_1_CTR_ADDRESS            0x4003
#define PAPU_PULSE_1_CTR_SIZE               IOREG_8
#define PAPU_PULSE_1_CTR_SHORT_DESCRIPTION  "pAPU_P_1_CTR"
#define PAPU_PULSE_1_CTR_COMMENT            "pAPU Pulse #1 Coarse Tune (CT) Register (W)"


// pulse 2
#define PAPU_PULSE_2_CR_ADDRESS             0x4004
#define PAPU_PULSE_2_CR_SIZE                IOREG_8
#define PAPU_PULSE_2_CR_SHORT_DESCRIPTION   "pAPU_P_2_CR"
#define PAPU_PULSE_2_CR_COMMENT             "pAPU Pulse #2 Control Register (W)"
                                            
#define PAPU_PULSE_2_RCR_ADDRESS            0x4005
#define PAPU_PULSE_2_RCR_SIZE               IOREG_8
#define PAPU_PULSE_2_RCR_SHORT_DESCRIPTION  "pAPU_P_2_RCR"
#define PAPU_PULSE_2_RCR_COMMENT            "pAPU Pulse #2 Ramp Control Register (W)"
                                            
#define PAPU_PULSE_2_FTR_ADDRESS            0x4006
#define PAPU_PULSE_2_FTR_SIZE               IOREG_8
#define PAPU_PULSE_2_FTR_SHORT_DESCRIPTION  "pAPU_P_2_FTR"
#define PAPU_PULSE_2_FTR_COMMENT            "pAPU Pulse #2 Fine Tune (FT) Register (W)"
                                            
#define PAPU_PULSE_2_CTR_ADDRESS            0x4007
#define PAPU_PULSE_2_CTR_SIZE               IOREG_8
#define PAPU_PULSE_2_CTR_SHORT_DESCRIPTION  "pAPU_P_2_CTR"
#define PAPU_PULSE_2_CTR_COMMENT            "pAPU Pulse #2 Coarse Tune (CT) Register (W)"


// triangle
#define PAPU_TRIANGLE_CR_1_ADDRESS           0x4008
#define PAPU_TRIANGLE_CR_1_SIZE              IOREG_8
#define PAPU_TRIANGLE_CR_1_SHORT_DESCRIPTION "pAPU_T_CR_1"
#define PAPU_TRIANGLE_CR_1_COMMENT           "pAPU Triangle Control Register #1 (W)"

#define PAPU_TRIANGLE_CR_2_ADDRESS           0x4009
#define PAPU_TRIANGLE_CR_2_SIZE              IOREG_8
#define PAPU_TRIANGLE_CR_2_SHORT_DESCRIPTION "pAPU_T_CR_2"
#define PAPU_TRIANGLE_CR_2_COMMENT           "pAPU Triangle Control Register #2 (?)"

#define PAPU_TRIANGLE_FR_1_ADDRESS           0x400A
#define PAPU_TRIANGLE_FR_1_SIZE              IOREG_8
#define PAPU_TRIANGLE_FR_1_SHORT_DESCRIPTION "pAPU_T_FR_1"
#define PAPU_TRIANGLE_FR_1_COMMENT           "pAPU Triangle Frequency Register #1 (W)"

#define PAPU_TRIANGLE_FR_2_ADDRESS           0x400B
#define PAPU_TRIANGLE_FR_2_SIZE              IOREG_8
#define PAPU_TRIANGLE_FR_2_SHORT_DESCRIPTION "pAPU_T_FR_2"
#define PAPU_TRIANGLE_FR_2_COMMENT           "pAPU Triangle Frequency Register #2 (W)"


// noise
#define PAPU_NOISE_CR_1_ADDRESS              0x400C
#define PAPU_NOISE_CR_1_SIZE                 IOREG_8
#define PAPU_NOISE_CR_1_SHORT_DESCRIPTION    "pAPU_N_CR_1"
#define PAPU_NOISE_CR_1_COMMENT              "pAPU Noise Control Register #1 (W)"

#define PAPU_NOISE_CR_2_ADDRESS              0x400D
#define PAPU_NOISE_CR_2_SIZE                 IOREG_8
#define PAPU_NOISE_CR_2_SHORT_DESCRIPTION    "Unused" //"pAPU_N_CR_2"
#define PAPU_NOISE_CR_2_COMMENT              "Unused Noise Control Register #2 ???" //"pAPU Noise Control Register #1 (W)"

#define PAPU_NOISE_FR_1_ADDRESS              0x400E
#define PAPU_NOISE_FR_1_SIZE                 IOREG_8
#define PAPU_NOISE_FR_1_SHORT_DESCRIPTION    "pAPU_N_FR_1"
#define PAPU_NOISE_FR_1_COMMENT              "pAPU Noise Frequency Register #1 (W)"

#define PAPU_NOISE_FR_2_ADDRESS              0x400F
#define PAPU_NOISE_FR_2_SIZE                 IOREG_8
#define PAPU_NOISE_FR_2_SHORT_DESCRIPTION    "pAPU_N_FR_2"
#define PAPU_NOISE_FR_2_COMMENT              "pAPU Noise Frequency Register #2 (W)"


// delta modulation
#define PAPU_DM_CR_ADDRESS                   0x4010
#define PAPU_DM_CR_SIZE                      IOREG_8
#define PAPU_DM_CR_SHORT_DESCRIPTION         "pAPU_DM_CR"
#define PAPU_DM_CR_COMMENT                   "pAPU Delta Modulation Control Register (W)"

#define PAPU_DM_DAR_ADDRESS                  0x4011
#define PAPU_DM_DAR_SIZE                     IOREG_8
#define PAPU_DM_DAR_SHORT_DESCRIPTION        "pAPU_DM_DAR"
#define PAPU_DM_DAR_COMMENT                  "pAPU Delta Modulation D/A Register (W)"

#define PAPU_DM_AR_ADDRESS                   0x4012
#define PAPU_DM_AR_SIZE                      IOREG_8
#define PAPU_DM_AR_SHORT_DESCRIPTION         "pAPU_DM_AR"
#define PAPU_DM_AR_COMMENT                   "pAPU Delta Modulation Address Register (W)"

#define PAPU_DM_DLR_ADDRESS                  0x4013
#define PAPU_DM_DLR_SIZE                     IOREG_8
#define PAPU_DM_DLR_SHORT_DESCRIPTION        "pAPU_DM_DLR"
#define PAPU_DM_DLR_COMMENT                  "pAPU Delta Modulation Data Length Register (W)"


// Clock Signal / Channel Control
#define PAPU_SV_CSR_ADDRESS                  0x4015
#define PAPU_SV_CSR_SIZE                     IOREG_8
#define PAPU_SV_CSR_SHORT_DESCRIPTION        "pAPU_SV_CSR"
#define PAPU_SV_CSR_COMMENT                  "pAPU Sound/Vertical Clock Signal Register (R)\n\n" \
                                             "   D6: Vertical Clock Signal IRQ Availability\n" \
                                             "   D4: Delta Modulation\n" \
                                             "   D3: Noise\n" \
                                             "   D2: Triangle\n" \
                                             "   D1: Pulse #2\n" \
                                             "   D0: Pulse #1\n" \
                                             "---------------------------------------------\n" \
                                             "pAPU Channel Control (W)\n\n" \
                                             "   D4: Delta Modulation\n" \
                                             "   D3: Noise\n" \
                                             "   D2: Triangle\n" \
                                             "   D1: Pulse #2\n" \
                                             "   D0: Pulse #1"




// Sprite DMA I/O regs

#define SPRITE_DMAR_ADDRESS                  0x4014
#define SPRITE_DMAR_SIZE                     IOREG_8
#define SPRITE_DMAR_SHORT_DESCRIPTION        "SPRITE_DMAR"
#define SPRITE_DMAR_COMMENT                  "Sprite DMA Register (W)\n\n" \
                                             "Transfers 256 bytes of memory into SPR-RAM. The address\n" \
                                             "read from is $100*N, where N is the value written."


// Joypad I/O regs

#define JOYPAD_1_ADDRESS                     0x4016
#define JOYPAD_1_SIZE                        IOREG_8
#define JOYPAD_1_SHORT_DESCRIPTION           "Joypad_1"
#define JOYPAD_1_COMMENT                     "Joypad #1 (RW)\n\n" \
                                             "READING:\n" \
                                             "   D4: Zapper Trigger\n" \
                                             "   D3: Zapper Sprite Detection\n" \
                                             "   D0: Joypad Data\n" \
                                             "----------------------------------------------\n" \
                                             "WRITING:\n" \
                                             "Joypad Strobe (W)\n\n" \
                                             "" \
                                             "   D0: Joypad Strobe\n" \
                                             "----------------------------------------------\n" \
                                             "WRITING:\n" \
                                             "Expansion Port Latch (W)\n" \
                                             "   D0: Expansion Port Method\n"

#define JOYPAD_2_ADDRESS                     0x4017
#define JOYPAD_2_SIZE                        IOREG_8
#define JOYPAD_2_SHORT_DESCRIPTION           "Joypad_2"
#define JOYPAD_2_COMMENT                     "Joypad #2/SOFTCLK (RW)\n\n" \
                                             "READING:\n" \
                                             "   D7: Vertical Clock Signal (External)\n" \
                                             "   D6: Vertical Clock Signal (Internal)\n" \
                                             "   D4: Zapper Trigger\n" \
                                             "   D3: Zapper Sprite Detection\n" \
                                             "   D0: Joypad Data\n" \
                                             "----------------------------------------------\n" \
                                             "WRITING:\n" \
                                             "Expansion Port Latch (W)\n\n" \
                                             "" \
                                             "   D0: Expansion Port Method\n"


#endif // _IOREGS_H

src/mappers.h
/*

	Nintendo Entertainment System (NES) loader module
	------------------------------------------------------
	Copyright 2006, Dennis Elser (dennis@backtrace.de)

*/


#ifndef _MAPPERS_H
#define _MAPPERS_H




#define MAPPER_NOT_SUPPORTED "mapper unknown/not supported"

char *mapper_names[] = {
    "no mapper used",
    "MMC 1",
    "UNROM",
    "CNROM",
    "MMC 3",
    "MMC 5",
    "FFE F4xxx",
    "AOROM",
    "FFE F3xxx",
    "MMC 2",
    "MMC 4",
    "Color Dreams",
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    "100 in 1",
    "Bandai",
    "FFE F8xxx",
    "Jaleco SS8806",
    "Namcot 106",
    MAPPER_NOT_SUPPORTED,
    "Konami VRC4",
    "Konami VRC2 Type A",
    "Konami VRC2 Type B",
    "Konami VRC6",
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    "Irem G 101",
    "Taito TC0190",
    "Nina 1",
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    "Tengen Rambo 1",
    "Irem H 3001",
    "GNROM",
    MAPPER_NOT_SUPPORTED,
    "Sunsoft Mapper 4",
    "Sunsoft FME7",
    MAPPER_NOT_SUPPORTED,
    "Camerica",
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    "Irem 74HC161 32",
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    MAPPER_NOT_SUPPORTED,
    "HK SF3"
};


enum
{
    MAPPER_NONE                     = 0,
    MAPPER_MMC1,
    MAPPER_UNROM,
    MAPPER_CNROM,
    MAPPER_MMC3,
    MAPPER_MMC5,
    MAPPER_FFE_F4XXX,
    MAPPER_AOROM,
    MAPPER_FFE_F3XXX,
    MAPPER_MMC2,
    MAPPER_MMC4,
    MAPPER_COLOR_DREAMS,

    MAPPER_100_IN_1                 = 15,
    MAPPER_BANDAI,
    MAPPER_FFE_F8XXX,
    MAPPER_JALECO_SS8806,
    MAPPER_NAMCOT_106,

    MAPPER_KONAMI_VRC4              = 21,
    MAPPER_KONAMI_VRC2_TYPE_A,
    MAPPER_KONAMI_VRC2_TYPE_B,
    MAPPER_KONAMI_VRC6,

    MAPPER_IREM_G_101               = 32,
    MAPPER_TAITO_TC0190,
    MAPPER_NINA_1,
    
    MAPPER_TENGEN_RAMBO_1           = 64,
    MAPPER_IREM_H_3001,
    MAPPER_GNROM,

    MAPPER_SUNSOFT_MAPPER_4         = 68,
    MAPPER_SUNSOFT_FME7,

    MAPPER_CAMERICA                 = 71,

    MAPPER_IREM_74HC161_32          = 78,

    MAPPER_HK_SF3                   = 91,


};

#define MAPPER_LAST                 91

#endif // _MAPPERS_H

src/nes.h
/*

	Nintendo Entertainment System (NES) loader module
	------------------------------------------------------
	Copyright 2006, Dennis Elser (dennis@backtrace.de)

*/


#ifndef _NES_H
#define _NES_H




//----------------------------------------------------------------------
//
//      general NES info:
//


#define RAM_START_ADDRESS                   0x0
#define RAM_SIZE                            0x2000

#define IOREGS_START_ADDRESS                0x2000
#define IOREGS_SIZE                         0x2020

#define EXPROM_START_ADDRESS                0x4020
#define EXPROM_SIZE                         0x1FE0

#define SRAM_START_ADDRESS                  0x6000
#define SRAM_SIZE                           0x2000

// start address and size of a trainer, if present
#define TRAINER_START_ADDRESS               0x7000
#define TRAINER_SIZE                        0x0200

#define ROM_START_ADDRESS                   0x8000
#define ROM_SIZE                            0x8000

#define PRG_PAGE_SIZE                       0x4000
#define CHR_PAGE_SIZE                       0x2000


#define PRG_ROM_BANK_SIZE                   PRG_PAGE_SIZE
#define PRG_ROM_8K_BANK_SIZE                0x2000
#define PRG_ROM_BANK_LOW_ADDRESS            ROM_START_ADDRESS
#define PRG_ROM_BANK_HIGH_ADDRESS           PRG_ROM_BANK_LOW_ADDRESS + PRG_ROM_BANK_SIZE
#define PRG_ROM_BANK_8000                   0x8000
#define PRG_ROM_BANK_A000                   0xA000
#define PRG_ROM_BANK_C000                   0xC000
#define PRG_ROM_BANK_E000                   0xE000


#define CHR_ROM_BANK_SIZE                   CHR_PAGE_SIZE
#define CHR_ROM_BANK_ADDRESS                RAM_START_ADDRESS



// start address of vectors
#define NMI_VECTOR_START_ADDRESS            0xFFFA
#define RESET_VECTOR_START_ADDRESS          0xFFFC
#define IRQ_VECTOR_START_ADDRESS            0xFFFE



// PPU RAM layout bottom-up

#define PATTERN_TABLE_SIZE                  0x1000
#define ATTRIBUTE_TABLE_SIZE                0x40
#define NAME_TABLE_SIZE                     0x3C0
#define MIRRORS_0_SIZE                      0xF00
#define MIRRORS_1_SIZE                      0xE0
#define MIRRORS_2_SIZE                      0xC000

#ifdef PALETTE_SIZE // defined in lines.hpp
    #undef PALETTE_SIZE
    #define PALETTE_SIZE                    0x10
#endif


#define PATTERN_TABLE_0_ADDRESS             0x0
#define PATTERN_TABLE_1_ADDRESS             0x1000

#define NAME_TABLE_0_ADDRESS                0x2000
#define ATTRIBUTE_TABLE_0_ADDRESS           0x23C0

#define NAME_TABLE_1_ADDRESS                0x2400
#define ATTRIBUTE_TABLE_1_ADDRESS           0x27C0

#define NAME_TABLE_2_ADDRESS                0x2800
#define ATTRIBUTE_TABLE_2_ADDRESS           0x2BC0

#define NAME_TABLE_3_ADDRESS                0x2C00
#define ATTRIBUTE_TABLE_3_ADDRESS           0x2CF0

#define MIRRORS_0_ADDRESS                   0x3000

#define IMAGE_PALETTE_ADDRESS               0x3F00

#define SPRITE_PALETTE_ADDRESS              0x3F10

#define MIRRORS_1_ADDRESS                   0x3F20

#define MIRRORS_2_ADDRESS                   0x4000



//----------------------------------------------------------------------
//
//      iNES file format specific information
//

// structure of iNES header
typedef struct _ines_hdr_t {

	uchar id[0x3];                          // NES
	uchar term;                             // 0x1A
	uchar prg_page_count_16k;               // number of PRG-ROM pages
	uchar chr_page_count_8k;                // number of CHR-ROM pages
	uchar rom_control_byte_0;               // flags describing ROM image
	uchar rom_control_byte_1;               // flags describing ROM image
	uchar ram_bank_count_8k;                // not used by this loader currently
	uchar reserved[7];                      // should all be zero (not checked by loader)

} ines_hdr;


// size of iNES header
#define INES_HDR_SIZE                       sizeof( ines_hdr )


// node name for iNES header
#define INES_HDR_NODE                       "$ iNES ROM header"

#define BANK_NUM_8000                       "$ Bank 8000"
#define BANK_NUM_C000                       "$ Bank C000"

// macros for masking control byte (cb) flags of the header
#define INES_MASK_V_MIRRORING( cb )         ( cb & 0x1 )
#define INES_MASK_H_MIRRORING( cb )         !INES_MASK_V_MIRRORING( cb )
#define INES_MASK_SRAM( cb )                ( (cb & 0x2) >> 1 )
#define INES_MASK_TRAINER( cb )             ( (cb & 0x4) >> 2 )
#define INES_MASK_VRAM_LAYOUT( cb )         ( (cb & 0x8) >> 3 )


// macro for getting the version of the mapper used by ROM image
#define INES_MASK_MAPPER_VERSION(cb0, cb1)  ( ((cb0 & 0xF0) >> 4) |  (cb1 & 0xF0) )




//----------------------------------------------------------------------
//
//      function prototypes for nes.cpp
//

static void load_ines_file( linput_t *li ); // convenience function for all below

static bool is_corrupt_ines_hdr( void );
static void fix_ines_hdr( void );

static void create_segments( linput_t *li ); // convenience function for the following few
static void create_sram_segment( void );
static void create_ram_segment( void );
static void create_ioreg_segment( void );
static void create_rom_segment( void );
static void create_exprom_segment( void );

static void load_trainer( linput_t *li );
static void load_chr_rom_bank( linput_t *li, uchar banknr, ea_t address );
static void load_prg_rom_bank( linput_t *li, uchar banknr, ea_t address );
static void load_8k_prg_rom_bank( linput_t *li, uchar banknr, ea_t address );
static void load_rom_banks( linput_t *li );

static void save_image_as_blobs( linput_t *li ); // convenience function for the following few
static bool save_ines_hdr_as_blob( void );
static bool save_trainer_as_blob( linput_t *li );
static bool save_prg_rom_pages_as_blobs( linput_t *li, uchar count );
static bool save_chr_rom_pages_as_blobs( linput_t *li, uchar count );


static char *get_mapper_name( uchar mapper );
static void define_item( ushort address, asize_t size, char *shortdesc, char *comment );
static ea_t get_vector( ea_t vec );
static void name_vector( ushort address, const char *name );
static bool add_entry_points( linput_t *li );
static void set_ida_export_data( void );
static void describe_rom_image( void );

#endif // _NES_H